### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Bender250**
New things for master:

* bugfix of false stream
* streams can be pipelined
* implementation of block ciphers
* more data for computation of final p-value
* fixed bugs with setting connector on and connector iteration

Consequence: change of config files, documented in wiki :).

### CODE REVIEW

File: streams/sha3/sha3_stream.cc [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
+hash_data(sha3_interface& hasher, const I& data, std::uint8_t* hash, const std::size_t hash_size) {
     using std::to_string;
 
-    int status = hasher.Init(int(8 * hash_size));
+    int status = hasher.Init(int(hash_size * 8));
     if (status != 0)
         throw std::runtime_error("cannot initialize hash (code: " + to_string(status) + ")");
 
-    status = hasher.Update(data.data(), 8 * data.size());
+    status = hasher.Update(&(*data.begin()), 8 * (data.end() - data.begin()));
```
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: You can do better than this 👎

File: streams/sha3/sha3_stream.cc [hunk 2]
```diff
@@ -21,28 +22,32 @@ hash_data(sha3_interface& hasher, const I& data, std::uint8_t* hash, std::size_t
...
     , _round(config.at("round"))
     , _hash_size(std::size_t(config.at("hash-bitsize")) / 8)
-    , _counter(8)
-    , _hasher(sha3_factory::create(_algorithm, unsigned(_round))) {
+    , _source(make_stream(config.at("source"), seeder, _hash_size)) // TODO: hash-input-size?
+    , _hasher(sha3_factory::create(_algorithm, unsigned(_round)))
+    , _data((_hash_size > osize)
+                    ? _hash_size
+                    : (_hash_size % osize) ? ((osize / _hash_size) + 1) * _hash_size
+                                           : osize) { // round osize to multiple of _hash_size
```
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: Please extract this nested ternary expression to a separate function.

File: streams/sha3/sha3_stream.cc [hunk 3]
```diff
@@ -21,28 +22,32 @@ hash_data(sha3_interface& hasher, const I& data, std::uint8_t* hash, std::size_t
         throw std::runtime_error("cannot finalize the hash (code: " + to_string(status) + ")");
 }
 
-sha3_stream::sha3_stream(const json& config)
-    : _algorithm(config.at("algorithm").get<std::string>())
+sha3_stream::sha3_stream(const json& config, default_seed_source &seeder, std::size_t osize)
+    : stream(osize)
+    , _algorithm(config.at("algorithm").get<std::string>())
```
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: You can remove this unnecessary attribute.

File: streams/sha3/sha3_stream.cc [hunk 4]
```diff
@@ -21,28 +22,32 @@ hash_data(sha3_interface& hasher, const I& data, std::uint8_t* hash, std::size_t
...
 
-    if ((set.rawsize() % _hash_size) != 0)
-        throw std::runtime_error("SHA-3 dataset size must be multiple of hash size");
+    for (auto beg = _data.begin(); beg != _data.end(); beg += _hash_size) {
+        vec_view view = _source->next();
 
-    for (; beg != end; beg += _hash_size) {
-        hash_data(*_hasher, _counter, beg, _hash_size);
-        _counter.increment();
+        hash_data(*_hasher, view, &(*beg), _hash_size);
```
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: You dont need to take pointer `&(*beg)`, just use simply `beg` instead.

File: streams/sha3/sha3_stream.h [hunk 5]
```diff
@@ -3,21 +3,24 @@
 #include <core/json.h>
 #include <core/stream.h>
 #include <memory>
+#include <../../core/random.h>
```
Comment (xtracko) [inline streams/sha3/sha3_stream.h]: `<../../core/random.h>` relative paths are forbidden.

File: streams/sha3/sha3_stream.h [hunk 6]
```diff
@@ -3,21 +3,24 @@
...
-    sha3_stream(const json& config);
+    sha3_stream(const json& config, default_seed_source& seeder, std::size_t osize);
     sha3_stream(sha3_stream&&);
     ~sha3_stream();
 
-    void read(dataset& set) override;
+    vec_view next() override;
 
 private:
     const std::string _algorithm;
```
Comment (xtracko) [inline streams/sha3/sha3_stream.h]: Remove this attribute

File: streams/sha3/sha3_stream.cc (hunk 1 above)
Comment (Bender250) [inline streams/sha3/sha3_stream.cc]: Very supportive, now I really know, how to solve this! I spent with this an hour to figure out I do not know how to do it better. Today again 15 minutes, I did not found better solution (I deleted this and after 15 minutes, I came again with this :/). So?

File: streams/sha3/sha3_stream.cc (hunk 4 above)
Comment (Bender250) [inline streams/sha3/sha3_stream.cc]: Somewhere I have to change it from iterator to raw data. How?

File: streams/sha3/sha3_stream.cc (hunk 1 above)
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: Simply, use std::distance ;)

File: streams/sha3/sha3_stream.cc (hunk 4 above)
Comment (xtracko) [inline streams/sha3/sha3_stream.cc]: std::vector::data() - iterate over raw data instead of iterator

Comment (Bender250) [inline streams/sha3/sha3_stream.cc]: And you would consider iteration over raw data as an better implementation?

```c
auto data = _data.data();
for (std::size_t i = 0; i < _data.size(); i += _hash_size) {
    vec_view view = _source->next();

    hash_data(*_hasher, view, &data[i], _hash_size);
}
```

File: streams/sha3/sha3_stream.cc (hunk 1 above)
Comment (Bender250) [inline streams/sha3/sha3_stream.cc]: Ahh, I though you dislike `&(*data.begin())`...

`view` already implements this as `size()`, done.

File: core/stream.h [hunk 7]
```diff
@@ -1,49 +1,25 @@
...
-#include <limits>
 #include <vector>
 
-struct counter {
-    using value_type = std::uint8_t;
-    using pointer = typename std::add_pointer<value_type>::type;
-    using const_pointer = typename std::add_pointer<const value_type>::type;
-    using iterator = std::vector<value_type>::iterator;
-    using const_iterator = std::vector<value_type>::const_iterator;
+using value_type = std::uint8_t;
```
Comment (Bender250) [inline core/stream.h]: I know, but I am lazy to fix it. I solve things like you do, someone else can solve my code.

**Pull request merged**